### PR TITLE
Poisson likelihood

### PR DESCRIPTION
**Pull request by frazane**
Adds the Poisson likelihood, along with an example notebook demonstrating a GP regression with count data.

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### CODE REVIEW

File: gpjax/likelihoods.py [hunk 1]
```diff
@@ -146,6 +145,33 @@ def predict(self, dist: tfd.Distribution) -> tfd.Distribution:
...
+        """The link function of the Poisson likelihood.
+
+        Args:
+            f (Float[Array, "N 1"]): Function values.
+
+        Returns:
+            tfd.Distribution: The likelihood function.
+        """
+        return tfd.Poisson(rate=jnp.exp(f))
+
```
Comment (Daniel-Dodd) [inline gpjax/likelihoods.py:160]: Looks good, but just a thought - is the `log_rate` argument more stable?

Comment (Daniel-Dodd) [reply gpjax/likelihoods.py:160]: After a brief look at the source code, the answer seems to be no.

**Review by thomaspinder — CHANGES_REQUESTED** [review]
Nice work! Two things that would be nice:
1) A test to check that a non-conjugate GP is returned from taking the product of the prior and Poisson likelihood.
2) A notebook demonstrating this. Simulated data would be fine and it could heavily draw on the MCMC routine used in the classification notebook.

File: examples/poisson.pct.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+#       format_version: '1.3'
+#       jupytext_version: 1.11.2
+#   kernelspec:
+#     display_name: base
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Count data regression with the Poisson likelihood
```
Comment (thomaspinder) [inline examples/poisson.pct.py:18]: Could we name this something short like "Count data regression"? It renders nicer in the docs' table of contents this way.

File: examples/poisson.pct.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+from jax.config import config
+from jaxtyping import Array, Float
+
+import gpjax as gpx
+
+# Enable Float64 for more stable matrix inversions.
+config.update("jax_enable_x64", True)
+tfd = tfp.distributions
+I = jnp.eye
+key = jr.PRNGKey(123)
```
Comment (thomaspinder) [inline examples/poisson.pct.py:44]: Could you please add the following to hook into Beartype?
```
with install_import_hook("gpjax", "beartype.beartype"):
    import gpjax as gpx
```

File: examples/poisson.pct.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+import tensorflow_probability.substrates.jax as tfp
+from jax.config import config
+from jaxtyping import Array, Float
+
+import gpjax as gpx
+
+# Enable Float64 for more stable matrix inversions.
+config.update("jax_enable_x64", True)
+tfd = tfp.distributions
+I = jnp.eye
```
Comment (thomaspinder) [inline examples/poisson.pct.py:43]: let's name this `identity_matrix` or simply use `jnp.eye`. Single letter variables are discouraged in PEP rules.

File: examples/poisson.pct.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+key = jr.PRNGKey(123)
+
+# %% [markdown]
+# ## Dataset
+#
+# For count data regression, the Poisson distribution is a natural choice. The probability mass function of the Poisson distribution is given by
```
Comment (thomaspinder) [inline examples/poisson.pct.py:49]: nit: _"is a natural choice for the likelihood function."_.

File: examples/poisson.pct.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+#
+# $$ p(y \,|\, \lambda) = \frac{\lambda^{y} e^{-\lambda}}{y!},$$
+#
+# where $y$ is the count and the parameter $\lambda$ is the rate of the Poisson distribution.
```
Comment (thomaspinder) [inline examples/poisson.pct.py:53]: nit: `\lambda \in \mathbb{R}_{>0}`

File: examples/poisson.pct.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+#
+# We than set $\lambda = \exp(f)$ where $f$ is the latent Gaussian process. The exponential function is the _link function_ for the Poisson distribution: it maps the output of a GP to the positive real line, which is suitable for modeling count data.
```
Comment (thomaspinder) [inline examples/poisson.pct.py:55]: Nice.

File: examples/poisson.pct.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    color="tab:blue",
+    label="95% CI",
+)
+
+# %% [markdown]
+# ## System configuration
+
+# %%
+# %load_ext watermark
+# %watermark -n -u -v -iv -w -a "Thomas Pinder & Daniel Dodd (edited by Francesco Zanetta)"
```
Comment (thomaspinder) [inline examples/poisson.pct.py:236]: You probably want to change - you wrote this notebook :)

**Review by thomaspinder — CHANGES_REQUESTED** [review]

**Review by thomaspinder — APPROVED** [review]

**Pull request merged**